src/main/scala/com/microsoft/hyperspace/index/rules/NestedRuleHelper.scala

```diff
@@ -0,0 +1,501 @@
+import org.apache.spark.sql.catalyst.optimizer.OptimizeIn
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, LeafNode, LogicalPlan, Project, Union}
+import org.apache.spark.sql.execution.datasources.InMemoryFileIndex
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
+import org.apache.spark.sql.types.{LongType, StructType}
+
+import com.microsoft.hyperspace.Hyperspace
+import com.microsoft.hyperspace.index.{FileInfo, IndexConstants, IndexLogEntry, IndexLogEntryTags}
+import com.microsoft.hyperspace.index.plans.logical.{BucketUnion, IndexHadoopFsRelation}
+import com.microsoft.hyperspace.index.rules.PlanUtils._
+import com.microsoft.hyperspace.util.{HyperspaceConf, ResolverUtils}
+
+class NestedRuleHelper(spark: SparkSession) extends BaseRuleHelper(spark) {
+
+  /**
+   * Transform the current plan to utilize index.
+   * The transformed plan reads data from indexes instead of the source relations.
+   * Bucketing information of the index is retained if useBucketSpec is true.
+   *
+   * NOTE: This method currently only supports transformation of nodes with supported relations.
+   *
+   * @param index Index used in transformation of plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Transformed logical plan that leverages an index.
+   */
+  override protected[rules] def transformPlanToUseIndexOnlyScan(
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    val provider = Hyperspace.getContext(spark).sourceProviderManager
+    // Note that depending on the case we transform only the base relation
+    // and sometimes other portions of the plan (e.g., filters). For instance,
+    // given the following input plan:
+    //        Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
+    // in the presence of a suitable index, we will transform to:
+    //        Project(A,B) -> Filter(C = 10) -> Index Scan (A,B,C)
+    // In the case of nested fields we will transform the project and
+    // filter nodes too.
+    plan transformUp {
+      case l: LeafNode if provider.isSupportedRelation(l) =>
+        val relation = provider.getRelation(l)
+        val location = index.withCachedTag(IndexLogEntryTags.INMEMORYFILEINDEX_INDEX_ONLY) {
+          new InMemoryFileIndex(spark, index.content.files, Map(), None)
+        }
+
+        val newSchema = StructType(
+          index.schema.filter(i => relation.plan.schema.exists(j => i.name.contains(j.name))))
+
+        val indexFsRelation = new IndexHadoopFsRelation(
+          location,
+          new StructType(),
+          newSchema,
+          if (useBucketSpec) Some(index.bucketSpec) else None,
+          new ParquetFileFormat,
+          Map(IndexConstants.INDEX_RELATION_IDENTIFIER))(spark, index)
+
+        val resolvedFields = ResolverUtils.resolve(
```

Comment:
Could you define some functions (outputFields, transformProject..) and override them rather than overriding whole transformPlanToUse*?
In this way, it's hard to maintain changes in transformPlanToUse*